validator.go

```diff
@@ -0,0 +1,201 @@
+	var now time.Time
+	vErr := new(ValidationError)
+
+	// Check, if we have a time func
+	if v.timeFunc != nil {
+		now = v.timeFunc()
+	} else {
+		now = time.Now()
+	}
+
+	if !v.VerifyExpiresAt(claims, now, false) {
+		exp := claims.GetExpirationTime()
+		delta := now.Sub(exp.Time)
+		vErr.Inner = fmt.Errorf("%s by %s", ErrTokenExpired, delta)
+		vErr.Errors |= ValidationErrorExpired
+	}
+
+	// Check iat if the option is enabled
+	if v.verifyIat && !v.VerifyIssuedAt(claims, now, false) {
+		vErr.Inner = ErrTokenUsedBeforeIssued
+		vErr.Errors |= ValidationErrorIssuedAt
+	}
+
+	if !v.VerifyNotBefore(claims, now, false) {
+		vErr.Inner = ErrTokenNotValidYet
+		vErr.Errors |= ValidationErrorNotValidYet
+	}
+
+	if v.expectedAud != "" && !v.VerifyAudience(claims, v.expectedAud, false) {
+		vErr.Inner = ErrTokenNotValidYet
+		vErr.Errors |= ValidationErrorNotValidYet
+	}
+
+	// Finally, we want to give the claim itself some possibility to do some
+	// additional custom validation based on their custom claims
+	cvt, ok := claims.(customValidationType)
+	if ok {
+		if err := cvt.CustomValidation(); err != nil {
+			vErr.Inner = err
+			vErr.Errors |= ValidationErrorClaimsInvalid
+		}
+	}
+
+	if vErr.valid() {
+		return nil
+	}
+
+	return vErr
+}
+
+// VerifyAudience compares the aud claim against cmp.
+// If required is false, this method will return true if the value matches or is unset
+func (v *Validator) VerifyAudience(claims Claims, cmp string, req bool) bool {
+	return verifyAud(claims.GetAudience(), cmp, req)
+}
+
+// VerifyExpiresAt compares the exp claim against cmp (cmp < exp).
+// If req is false, it will return true, if exp is unset.
+func (v *Validator) VerifyExpiresAt(claims Claims, cmp time.Time, req bool) bool {
+	exp := claims.GetExpirationTime()
+	if exp == nil {
+		return verifyExp(nil, cmp, req, v.leeway)
+	}
+
+	return verifyExp(&exp.Time, cmp, req, v.leeway)
+}
+
+// VerifyIssuedAt compares the iat claim against cmp (cmp >= iat).
+// If req is false, it will return true, if iat is unset.
+func (v *Validator) VerifyIssuedAt(claims Claims, cmp time.Time, req bool) bool {
+	iat := claims.GetIssuedAt()
+	if iat == nil {
+		return verifyIat(nil, cmp, req, v.leeway)
+	}
+
+	return verifyIat(&iat.Time, cmp, req, v.leeway)
+}
+
+// VerifyNotBefore compares the nbf claim against cmp (cmp >= nbf).
+// If req is false, it will return true, if nbf is unset.
+func (v *Validator) VerifyNotBefore(claims Claims, cmp time.Time, req bool) bool {
+	nbf := claims.GetNotBefore()
+	if nbf == nil {
+		return verifyNbf(nil, cmp, req, v.leeway)
+	}
+
+	return verifyNbf(&nbf.Time, cmp, req, v.leeway)
+}
+
+// VerifyIssuer compares the iss claim against cmp.
+// If required is false, this method will return true if the value matches or is unset
+func (v *Validator) VerifyIssuer(claims Claims, cmp string, req bool) bool {
+	return verifyIss(claims.GetIssuer(), cmp, req)
+}
+
+// ----- helpers
+
+func verifyAud(aud []string, cmp string, required bool) bool {
```

Comment:
It seems to me this function is copied from another file. If that's the case, I know you didn't write this code, but how do you feel about making a change based on my below statement?

I don't find the `subtle.ConstantTimeCompare([]byte(a), []byte(cmp)) != 0` favorable over a string comparison, but perhaps I don't fully understand its purpose.

Reply:
Puh, good question. The general idea is that all things "secret" should be compared in constant time to not leak any information about them via side channels (such as time). One could argue that the audience is not really secret, but I would keep this as is for now.